### PR TITLE
VEIOSS-241 Allow Excluded Traders to Re-join the IOSS Scheme

### PR DESCRIPTION
**Pull request by pbyrne84**

### CODE REVIEW

File: app/connectors/RegistrationHttpParser.scala [hunk 1]
```diff
@@ -29,7 +30,7 @@ object RegistrationHttpParser extends Logging {
 
   type RegistrationResultResponse = Either[ErrorResponse, EtmpEnrolmentResponse]
   type DisplayRegistrationResponse = Either[ErrorResponse, RegistrationWrapper]
-  type AmendRegistrationResultResponse = Either[ErrorResponse, Any]
+  type AmendRegistrationResultResponse = Either[ErrorResponse, AmendRegistrationResponse]
```
Comment (pbyrne84) [inline app/connectors/RegistrationHttpParser.scala:33]: Need the response back for the new IOSSNumber

File: app/controllers/BusinessContactDetailsController.scala [hunk 2]
```diff
@@ -78,13 +78,7 @@ class BusinessContactDetailsController @Inject()(
...
           }
 
-          val continueUrl = if (waypoints.inAmend) {
-            controllers.amend.routes.ChangeRegistrationController.onPageLoad().url
-          } else if (waypoints.currentMode == CheckMode) {
-            routes.CheckYourAnswersController.onPageLoad().url
-          } else {
-            routes.BankDetailsController.onPageLoad(waypoints).url
-          }
+          val continueUrl = waypoints.getNextCheckYourAnswersPageFromWaypoints.getOrElse(BankDetailsPage).route(waypoints).url
```
Comment (pbyrne84) [inline app/controllers/BusinessContactDetailsController.scala:81]:
```scala
 def getNextCheckYourAnswersPageFromWaypoints: Option[CheckAnswersPage] = {
      waypoints match {
        case nonEmptyWaypoints: NonEmptyWaypoints =>
          List(RejoinRegistrationPage, ChangeRegistrationPage, CheckYourAnswersPage).find { page =>
            nonEmptyWaypoints.waypoints.toList.map(_.urlFragment).contains(page.urlFragment)
          }

        case _ =>
          None
      }
    }
  }
```

Just prioritises across waypoints, all journey tests work

File: app/controllers/tradingNames/HasTradingNameController.scala [hunk 3]
```diff
@@ -44,41 +44,24 @@ class HasTradingNameController @Inject()(
...
+              case None => form
+              case Some(value) => form.fill(value)
+            }
+
+            Ok(view(preparedForm, waypoints, companyName)).toFuture
+        }
+    }
   }
 
   private def getCompanyName(waypoints: Waypoints)(block: String => Future[Result])
```
Comment (pbyrne84) [inline app/controllers/tradingNames/HasTradingNameController.scala:63]: Just moved method up as it is shared between both methods

File: app/models/etmp/amend/AmendRegistrationResponse.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+
+package models.etmp.amend
+
+import play.api.libs.json.{Json, OFormat}
+
+import java.time.LocalDateTime
+
+case class AmendRegistrationResponse(
```
Comment (pbyrne84) [inline app/models/etmp/amend/AmendRegistrationResponse.scala:23]: Now returned from the backend as we need the new IOSSNumber

File: app/pages/BusinessContactDetailsPage.scala [hunk 5]
```diff
@@ -37,8 +37,7 @@ case object BusinessContactDetailsPage extends QuestionPage[BusinessContactDetai
 
   override protected def nextPageCheckMode(waypoints: NonEmptyWaypoints, answers: UserAnswers): Page =
     answers.get(this) match {
-      case Some(_) if waypoints.inAmend => ChangeRegistrationPage
-      case Some(_) => CheckYourAnswersPage
+      case Some(_) => waypoints.getNextCheckYourAnswersPageFromWaypoints.getOrElse(CheckYourAnswersPage)
```
Comment (pbyrne84) [inline app/pages/BusinessContactDetailsPage.scala:40]: Handle rejoin as well

File: app/models/etmp/EtmpDisplayRegistration.scala [hunk 6]
```diff
@@ -18,13 +18,44 @@ package models.etmp
...
+      val minimumDate = currentDate.minusYears(2)
+      etmpExclusion.effectiveDate.isBefore(minimumDate)
+    } else {
+      false
+    }
+  }
+
+  private def notQuarantinedAndAfterEffectiveDate(currentDate: LocalDate, etmpExclusion: EtmpExclusion): Boolean = {
+    if(!etmpExclusion.quarantine) {
+      etmpExclusion.effectiveDate.isBefore(currentDate) || etmpExclusion.effectiveDate.isEqual(currentDate)
```
Comment (hmrcvmu) [inline app/models/etmp/EtmpDisplayRegistration.scala:52]: Could add this file https://github.com/hmrc/ioss-exclusions-frontend/blob/0d195d3ada8cd70cf231238ca0bb33567c274ece/app/date/package.scala to the project and use `etmpExclusion.effectiveDate <= currentDate`

Comment (pbyrne84) [reply app/models/etmp/EtmpDisplayRegistration.scala]: I also changed this in the other front end. Joy of copy paste coding.

https://github.com/hmrc/ioss-returns-frontend/blob/181729a186cf397336f8004b9a2d81f667d5e605/app/models/etmp/EtmpDisplayRegistration.scala#L52

File: app/services/RegistrationService.scala [hunk 7]
```diff
@@ -56,19 +56,28 @@ class RegistrationService @Inject()(
     registrationConnector.createRegistration(buildEtmpRegistrationRequest(answers, vrn, commencementDate))
   }
 
+
+  def getRegistration()(implicit hc: HeaderCarrier): Future[DisplayRegistrationResponse] = {
+    registrationConnector.getRegistration()
+  }
```
Comment (DevAndyroo) [inline app/services/RegistrationService.scala:62]: what's the point of this function?

**Review by DevAndyroo — CHANGES_REQUESTED** [review]

File: app/services/RegistrationService.scala (hunk 7 above)
Comment (pbyrne84) [reply app/services/RegistrationService.scala]: Dead code, was using it, but then wasn't. Have deleted it

**Review by DevAndyroo — APPROVED** [review]

**Pull request merged**